Firebase/Messaging/FIRMessagingExtensionHelper.m

```diff
@@ -0,0 +1,118 @@
+
+  NSString *currentImageURL = content.userInfo[kPayloadOptionsName][kPayloadOptionsImageURLName];
+  if (!currentImageURL) {
+    [self deliverNotification];
+    return;
+  }
+#if TARGET_OS_IOS
+  NSURL *attachmentURL = [NSURL URLWithString:currentImageURL];
+  if (attachmentURL) {
+    [self loadAttachmentForURL:attachmentURL
+             completionHandler:^(UNNotificationAttachment *attachment) {
+               self.bestAttemptContent.attachments = @[ attachment ];
+               [self deliverNotification];
+             }];
+  } else {
+    FIRMessagingLoggerError(kFIRMessagingServiceExtensionImageInvalidURL,
+                            @"The Image URL provided is invalid %@.", currentImageURL);
+    [self deliverNotification];
+  }
+#else
+  [self deliverNotification];
+#endif
+}
+
+#if TARGET_OS_IOS
+- (void)loadAttachmentForURL:(NSURL *)attachmentURL
+           completionHandler:(void (^)(UNNotificationAttachment *))completionHandler NS_AVAILABLE_IOS(10.0) {
+  __block UNNotificationAttachment *attachment = nil;
+
+  NSURLSession *session = [NSURLSession
+      sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
+  [[session
+      downloadTaskWithURL:attachmentURL
+        completionHandler:^(NSURL *temporaryFileLocation, NSURLResponse *response, NSError *error) {
+          if (error != nil) {
+            FIRMessagingLoggerError(kFIRMessagingServiceExtensionImageNotDownloaded,
+                                    @"Failed to download image given URL %@, error: %@\n",
+                                    attachmentURL, error);
+            completionHandler(attachment);
+            return;
+          }
+
+          NSFileManager *fileManager = [NSFileManager defaultManager];
+          NSString *fileExtension =
+              [NSString stringWithFormat:@".%@", [response.suggestedFilename pathExtension]];
+          NSURL *localURL = [NSURL
+              fileURLWithPath:[temporaryFileLocation.path stringByAppendingString:fileExtension]];
+          [fileManager moveItemAtURL:temporaryFileLocation toURL:localURL error:&error];
+          if (error) {
+            FIRMessagingLoggerError(
+                kFIRMessagingServiceExtensionLocalFileNotCreated,
+                @"Failed to move the image file to local location: %@, error: %@\n", localURL,
+                error);
+            completionHandler(attachment);
+            return;
+          }
+
+          attachment = [UNNotificationAttachment attachmentWithIdentifier:@""
+                                                                      URL:localURL
+                                                                  options:nil
+                                                                    error:&error];
+          if (error) {
+            FIRMessagingLoggerError(kFIRMessagingServiceExtensionImageNotAttached,
+                                    @"Failed to attach the image to attachments, error: %@\n",
```

Comment:
Should message be "Failed to create attachment"?  Should the localURL be part of the error message?

Reply:
Good call.